### ISSUE TITLE
Able to create a v1alpha3 machine without ClusterName by using v1alpha2 yaml

### ISSUE THREAD

**Issue by sedefsavas** [kubernetes-sigs/cluster-api#2518]
**What steps did you take and what happened:**
Applied the following v1alpha2 yaml to v1alpha3 controller. This yaml does not have a "cluster.x-k8s.io/cluster-name" label and "cluster.x-k8s.io/conversion-data" annotation which is supposed to hold v1alpha3 data is missing ClusterName too and it has an additional field that v1alpha3 Machine Spec does not have. Machine is created with clusterName: "".

```
apiVersion: cluster.x-k8s.io/v1alpha2
kind: Machine
metadata:
  labels:
    cluster.x-k8s.io/control-plane: "true"
  annotations:
    cluster.x-k8s.io/conversion-data: "{\"metadata\":{\"name\":\"test-1\",\"creationTimestamp\":null,\"labels\":{\"label1\":\"\"}},\"spec\":{\"notClusterName\":\"cname\",\"bootstrap\":{},\"infrastructureRef\":{}},\"status\":{\"bootstrapReady\":true,\"infrastructureReady\":true}}"
  name: management-clusterx-controlplane-0
  namespace: default
spec:
  bootstrap:
    configRef:
      apiVersion: bootstrap.cluster.x-k8s.io/v1alpha2
      kind: KubeadmConfig
      name: management-clusterx-controlplane-0
      namespace: default
  infrastructureRef:
    apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
    kind: VSphereMachine
    name: management-clusterx-controlplane-0
    namespace: default
  version: 1.16.2
```

**What did you expect to happen:**
I expected it to be failed during creation.

**Environment:**

- Cluster-api version:  v0.3.0-rc.2

/kind bug

**Comment by vincepri** [comment]
Looked at this issue with @sedefsavas a little ago, it seems like that regardless of the annotation, we're able to get around OpenAPI validation defined on the stored version (v1alpha3). 


Result when reading back the object:
```yaml
apiVersion: cluster.x-k8s.io/v1alpha3
kind: Machine
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"cluster.x-k8s.io/v1alpha2","kind":"Machine","metadata":{"annotations":{},"labels":{"cluster.x-k8s.io/control-plane":"true"},"name":"management-clusterx-controlplane-0","namespace":"default"},"spec":{"bootstrap":{"configRef":{"apiVersion":"bootstrap.cluster.x-k8s.io/v1alpha2","kind":"KubeadmConfig","name":"management-clusterx-controlplane-0","namespace":"default"}},"infrastructureRef":{"apiVersion":"infrastructure.cluster.x-k8s.io/v1alpha2","kind":"VSphereMachine","name":"management-clusterx-controlplane-0","namespace":"default"},"version":"1.16.2"}}
  creationTimestamp: "2020-03-03T23:51:34Z"
  generation: 1
  labels:
    cluster.x-k8s.io/control-plane: "true"
  name: management-clusterx-controlplane-0
  namespace: default
  resourceVersion: "290843"
  selfLink: /apis/cluster.x-k8s.io/v1alpha3/namespaces/default/machines/management-clusterx-controlplane-0
  uid: d376a0bd-f50e-4f70-bebf-2c03aaf7862a
spec:
  bootstrap:
    configRef:
      apiVersion: bootstrap.cluster.x-k8s.io/v1alpha2
      kind: KubeadmConfig
      name: management-clusterx-controlplane-0
      namespace: default
  clusterName: ""
  infrastructureRef:
    apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
    kind: VSphereMachine
    name: management-clusterx-controlplane-0
    namespace: default
  version: 1.16.2
status:
  bootstrapReady: false
  infrastructureReady: false
```

The `clusterName: ""` is the interesting piece, in v1alpha3 we added the `ClusterName` field as required, it also has `// +kubebuilder:validation:MinLength=1` as marker.

If I create the Machine v1alpha3 yaml above (removing some metadata field) on a new cluster I get the error `The Machine "management-clusterx-controlplane-0" is invalid: spec.clusterName: Invalid value: "": spec.clusterName in body should be at least 1 chars long`

@liggitt do you have any pointers/suggestions on this issue?

**Comment by liggitt** [comment]
are you asking why v1alpha3 validation doesn't run when submitting a v1alpha2 request?

**Comment by liggitt** [comment]
If you want to add a new field in v1alpha3 AND make it required, you have to ensure every existing v1alpha2 object converts to a valid v1alpha3 object by providing a default value for that field in conversion

**Comment by vincepri** [comment]
🙂 good question! Things got complicated here because we usually assume the field can be converted from a label. The history behind `clusterName` is that it was a label before, it has been "required" as in the controller would error out if the field wasn't set. 

The assumption I was making might be wrong, but I was expecting the following:
1. Submit a v1alpha2 object, the label isn't set,  object is sent to the apiserver
2. API server sees v1alpha2, calls v1alpha3 webhooks
3. Webhooks do their thing (`spec.clusterName` is still an empty string), object is sent back 
4. API server receives v1alpha3 version, applies OpenAPI validation and refuses to store it
5. Error is sent back

I guess the 4th point isn't what's happening, and we need to find another way to make throw an error?

**Comment by liggitt** [comment]
I'm mobile at the moment, but I think conversion to the storage version happens post validation (validation was historically concerned with ensuring good input from the user and conversion was assumed to be a mechanical known-good operation)

**Comment by vincepri** [comment]
Thank you, if that's the assumption I can think of some solutions:

- Add a check to require the `cluster-name` label in validation webhooks — once we move to `MatchPolicy: Equivalent` this should also for v1alpha2 converted object and return an error.
- Add a check to controllers and error out (less preferred, but definitely a quicker fix).

**Comment by ncdc** [comment]
@vincepri @sedefsavas we do not currently have webhook validation logic that requires spec.clusterName. If you add that, and then try to create using the yaml above, you get this:

```
Error from server (Machine.cluster.x-k8s.io "management-clusterx-controlplane-0" is invalid: spec.clusterName: Required value: ANDY): error when creating "STDIN": admission webhook "validation.machine.cluster.x-k8s.io" denied the request: Machine.cluster.x-k8s.io "management-clusterx-controlplane-0" is invalid: spec.clusterName: Required value: (your error here 😄)
```

**Comment by vincepri** [comment]
Are we going to end up having to duplicate OpenAPI validations in webhooks?

**Comment by vincepri** [comment]
Bumping investigation / possible fixes to v0.3.x

/milestone v0.3.x

**Comment by vincepri** [comment]
/close

Closing given that we won't fix this in v0.3.x.
